File: src/main/java/io/github/stiv3ns/twactionorganizer/pmsender/PMFormatter.java
package io.github.stiv3ns.twactionorganizer.pmsender;

import io.github.stiv3ns.twactionorganizer.twao.Player;
import io.github.stiv3ns.twactionorganizer.twao.VillageAssignment;
import io.github.stiv3ns.twactionorganizer.twao.World;
import io.github.stiv3ns.twactionorganizer.twao.villages.Village;

import java.text.SimpleDateFormat;
import java.util.*;
import java.util.List;

public class PMFormatter {
    private final World         world;
    private final Calendar      dateOfArrival;

    private static final int    SECONDS_PER_MIN        = 60;
    private static final int    GROUP_SIZE             = 5;
    private static final String OPEN_SPOILER        = "[spoiler]\n";
    private static final String CLOSE_SPOILER       = "[/spoiler]\n\n\n";

    private static ResourceBundle bundle = ResourceBundle.getBundle("localization/pmformatter");

    private static String REQUIREMENTS_HEADER = bundle.getString("REQUIREMENTS_HEADER");
    private static String NOBLE_HEADER        = bundle.getString("NOBLE_HEADER");
    private static String OFF_HEADER          = bundle.getString("OFF_HEADER");
    private static String FAKE_HEADER         = bundle.getString("FAKE_HEADER");
    private static String FAKENOBLE_HEADER    = bundle.getString("FAKENOBLE_HEADER");
    private static String EXECUTION_TEXT      = bundle.getString("EXECUTION_TEXT");

    private enum Unit {
        RAM,
        NOBLE
    }

    private int previousDay = -1; //flag init
    private boolean nextDay = false; //flag init

    /**
     * @param world             Specific world to get information from (e.g. speed)
     * @param dateOfArrival     Concrete date that is meant to be "D-Day"
     */
    public PMFormatter(World world, Calendar dateOfArrival) {
        this.world = world;
        this.dateOfArrival = dateOfArrival;
    }

    /**
     * Generates a message intended for a given player.
     *
     * @param player    Given player
     * @return          Message with all assignments
     */
    public String generatePlayerMsg(Player player) {
        StringBuilder sbuilder = new StringBuilder();

        if (player.getNobleAssignmentsCopy().size() > 0 || player.getFakeNobleAssignmentsCopy().size() > 0) {
            sbuilder.append(REQUIREMENTS_HEADER);
            generateNobleRequirements(sbuilder, player);
        }

        if (player.getNobleAssignmentsCopy().size() > 0) {
            sortAssignmentsList(player.getNobleAssignmentsCopy());

            sbuilder.append(NOBLE_HEADER);
            generateCommandsList(sbuilder, player.getNobleAssignmentsCopy(), getUnitSpeed(Unit.NOBLE));
        }

        if (player.getOffAssignmentsCopy().size() > 0) {
            sortAssignmentsList(player.getOffAssignmentsCopy());

            sbuilder.append(OFF_HEADER);
            generateCommandsList(sbuilder, player.getOffAssignmentsCopy(), getUnitSpeed(Unit.RAM));
        }

        if (player.getFakeAssignmentsCopy().size() > 0) {
            sortAssignmentsList(player.getFakeAssignmentsCopy());

            sbuilder.append(FAKE_HEADER);
            generateCommandsList(sbuilder, player.getFakeAssignmentsCopy(), getUnitSpeed(Unit.RAM));
        }

        if (player.getFakeNobleAssignmentsCopy().size() > 0) {
            sortAssignmentsList(player.getFakeNobleAssignmentsCopy());

            sbuilder.append(FAKENOBLE_HEADER);
            generateCommandsList(sbuilder, player.getFakeNobleAssignmentsCopy(), getUnitSpeed(Unit.NOBLE));
        }

        return sbuilder.toString();
    }

    private void generateNobleRequirements(StringBuilder sbuilder, Player player) {
        HashMap<Village, Integer> requirements = new HashMap<>();

        List<VillageAssignment> nobleAssignments = new LinkedList<>();
        nobleAssignments.addAll(player.getNobleAssignmentsCopy());
        nobleAssignments.addAll(player.getFakeNobleAssignmentsCopy());

        sbuilder.append(OPEN_SPOILER);

        int previousValue;
        for (VillageAssignment a : nobleAssignments) {
            if (requirements.get(a.getDeparture()) == null) {
                requirements.put(a.getDeparture(), 1);
            } else {
                previousValue = requirements.get(a.getDeparture());
                requirements.put(a.getDeparture(), ++previousValue);
            }
        }

        // {1: [Village, Village], 2: [], 3: [], ... }
        HashMap<Integer, List<Village>> sortedMap = sortRequirements(requirements);

        int iterator = 0;
        for (int value : sortedMap.keySet()) {
            for (Village vil : sortedMap.get(value)) {
                sbuilder.append(String.format("%d || %s\n", value, vil.toString()));

                if (++iterator % GROUP_SIZE == 0) {
                    sbuilder.append("\n\n");
                }
            }
        }

        sbuilder.append(CLOSE_SPOILER);
    }

    private void generateCommandsList(StringBuilder sbuilder, List<VillageAssignment> assignmentsList, int speed) {
        String departureTime, possibleNewLine, assignment, domain;
        int counter = 0, departureId, destinationId;

        sbuilder.append(OPEN_SPOILER);

        for (VillageAssignment a : assignmentsList) {
            departureTime = getDepartureTime(a.getSquaredDistance(), speed);

            possibleNewLine = nextDay ? "\n\n\n" : "";
            assignment = a.toString();
            domain = world.getDomain();
            departureId = a.getDeparture().getId();
            destinationId = a.getDestination().getId();

            sbuilder.append( String.format("%s%d. %s %s\n", possibleNewLine, ++counter, departureTime, assignment) );

            sbuilder.append( String.format("[url=%s/game.php?village=%d&screen=place&target=%d]%s[/url]\n", domain, departureId, destinationId, EXECUTION_TEXT) );
        }

        sbuilder.append(CLOSE_SPOILER);
    }
    
    private String getDepartureTime(double squaredDistance, int standardTimePerUnit) {

        Calendar departureTime = (Calendar) dateOfArrival.clone();

        double distance = Math.sqrt(squaredDistance);
        double unitSpeed = standardTimePerUnit / world.getSpeed();
        int travelTime = (int) ((distance * unitSpeed) * SECONDS_PER_MIN );

        departureTime.add(Calendar.SECOND, -travelTime);

        int dayOfMonth = departureTime.get(Calendar.DAY_OF_MONTH);

        if (dayOfMonth != previousDay) {
            nextDay = true;
        } else {
            nextDay = false;
        }
        previousDay = dayOfMonth;

        return new SimpleDateFormat("dd.MM | HH:mm:ss").format(departureTime.getTime());
    }

    private void sortAssignmentsList(List<VillageAssignment> list) {
        list.sort(Comparator.comparing(VillageAssignment::getSquaredDistance));
        Collections.reverse(list);
    }

    private HashMap<Integer, List<Village>> sortRequirements(HashMap<Village, Integer> map) {
        HashMap<Integer, List<Village>> sortedMap = new HashMap<>();

        int value;
        List<Village> vList;
        for (Village vil : map.keySet()) {
            value = map.get(vil);
            if (sortedMap.get(value) == null) {
                vList = new LinkedList<>();
                vList.add(vil);
                sortedMap.put(value, vList);
            } else {
                sortedMap.get(value).add(vil);
            }
        }

        return sortedMap;
    }

    private int getUnitSpeed(Unit type) {
        int speed;
        switch (type) {
            case RAM:
                speed = 30;
                break;
            case NOBLE:
                speed = 35;
                break;
            default:
                speed = 30;
                break;
        }
        return speed;
    }
}

File: src/main/java/io/github/stiv3ns/twactionorganizer/pmsender/PMFormatter.kt
package io.github.stiv3ns.twactionorganizer.pmsender

import io.github.stiv3ns.twactionorganizer.twao.Player
import io.github.stiv3ns.twactionorganizer.twao.VillageAssignment
import io.github.stiv3ns.twactionorganizer.twao.World
import io.github.stiv3ns.twactionorganizer.twao.villages.Village
import java.lang.StringBuilder
import java.time.LocalDateTime
import java.time.format.DateTimeFormatter
import java.util.*
import kotlin.math.sqrt

class PMFormatter(private val world: World, private val dateOfArrival: LocalDateTime) {

    private val OPEN_SPOILER = "[spoiler]\n"
    private val CLOSE_SPOILER = "[/spoiler]\n\n\n"

    private val bundle = ResourceBundle.getBundle("localization/pmformatter")

    private val REQUIREMENTS_HEADER by lazy { bundle.getString("REQUIREMENTS_HEADER") }
    private val NOBLE_HEADER        by lazy { bundle.getString("NOBLE_HEADER") }
    private val OFF_HEADER          by lazy { bundle.getString("OFF_HEADER") }
    private val FAKE_HEADER         by lazy { bundle.getString("FAKE_HEADER") }
    private val FAKENOBLE_HEADER    by lazy { bundle.getString("FAKENOBLE_HEADER") }
    private val EXECUTION_TEXT      by lazy { bundle.getString("EXECUTION_TEXT") }

    private val GROUP_SIZE = 5

    private enum class TroopType {
        RAM,
        NOBLE
    }

    fun getFormattedMsgFor(player: Player): String {
        val msg = StringBuilder()

        if (player.nobleAssignmentsCopy.isNotEmpty() || player.fakeNobleAssignmentsCopy.isNotEmpty()) {
            msg.append(REQUIREMENTS_HEADER)
            appendNobleRequirements(msg, player)
        }

        if (player.nobleAssignmentsCopy.isNotEmpty()) {
            msg.append(NOBLE_HEADER)
            appendAssignments(msg, player.nobleAssignmentsCopy, TroopType.NOBLE)
        }

        if (player.offAssignmentsCopy.isNotEmpty()) {
            msg.append(OFF_HEADER)
            appendAssignments(msg, player.offAssignmentsCopy, TroopType.RAM)
        }

        if (player.fakeNobleAssignmentsCopy.isNotEmpty()) {
            msg.append(FAKENOBLE_HEADER)
            appendAssignments(msg, player.fakeNobleAssignmentsCopy, TroopType.NOBLE)
        }

        if (player.fakeAssignmentsCopy.isNotEmpty()) {
            msg.append(FAKE_HEADER)
            appendAssignments(msg, player.fakeAssignmentsCopy, TroopType.RAM)
        }

        return msg.toString()
    }

    private fun appendNobleRequirements(msg: StringBuilder, player: Player) {
        val allNobleAssignments = player.nobleAssignmentsCopy + player.fakeNobleAssignmentsCopy

        val requirementsByVillage = numberOfNobleRequirementsByVillage(allNobleAssignments)
        val sortedRequirements = sortedRequirementsByVillage(requirementsByVillage)

        msg.append(OPEN_SPOILER)

        var iteratorCounter = 0
        for ((village, numberOfRequiredNobles) in sortedRequirements) {
            msg.append("$numberOfRequiredNobles __ $village\n")

            if (++iteratorCounter % GROUP_SIZE == 0) {
                msg.append("\n")
            }
        }

        msg.append(CLOSE_SPOILER)
    }

    private fun appendAssignments(msg: StringBuilder, assignments: List<VillageAssignment>, slowestTroop: TroopType) {
        msg.append(OPEN_SPOILER)

        val sortedAssignments = sortedAssignments(assignments)
        var counter = 0
        var previousDayOfMonth = -1
        for (assignment in sortedAssignments) {
            val departureTime = designateDepartureTime(assignment.squaredDistance, slowestTroop)

            val verticalSpaceIfNextDay = if (previousDayOfMonth != departureTime.dayOfMonth) "\n\n\n" else ""
            val verticalSpaceIfNextGroup = if (counter % GROUP_SIZE == 0) "\n" else ""

            msg.append("$verticalSpaceIfNextDay ${++counter}. ${formatDepartureTime(departureTime)} $assignment \n")
               .append("[url=${world.domain}/game.php")
               .append("?village=${assignment.departure.id ?: 0}")
               .append("&screen=place&target=${assignment.destination.id ?: 0}]")
               .append("$EXECUTION_TEXT[/url]\n")
               .append("$verticalSpaceIfNextGroup")

            previousDayOfMonth = departureTime.dayOfMonth
        }

        msg.append(CLOSE_SPOILER)
    }

    private fun numberOfNobleRequirementsByVillage(allNobleAssignments: List<VillageAssignment>): Map<Village, Int> {
        val assignmentsByDeparture = allNobleAssignments.groupBy { it.departure }
        val requirementsByVillage = assignmentsByDeparture.mapValues { it.value.size }

        return requirementsByVillage
    }

    private fun sortedRequirementsByVillage(requirementsByVillage: Map<Village, Int>): SortedMap<Village, Int>
        = requirementsByVillage.toSortedMap( compareBy { requirementsByVillage[it] } ) /* sort by values */

    private fun designateDepartureTime(squaredDistance: Int, slowestTroop: TroopType): LocalDateTime {
        val actualDistance = sqrt(squaredDistance.toDouble())
        val travelTime = actualDistance * troopTravelTime(slowestTroop)

        val departureTime = dateOfArrival.minusMinutes(travelTime.toLong())

        return departureTime
    }

    private fun formatDepartureTime(departureTime: LocalDateTime): String
        = departureTime.format(DateTimeFormatter.ofPattern("dd/MM/YY | HH:mm:ss"))

    private fun sortedAssignments(assignments: List<VillageAssignment>)
        = assignments.sortedByDescending { it.squaredDistance }

    private fun troopTravelTime(type: TroopType) = when (type) {
        TroopType.RAM -> 30
        TroopType.NOBLE -> 35
    } / world.speed
}